File: transaction-analyzer/scala/src/main/scala/InputStreamTransactionLoader.scala

import java.io.InputStream
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import scala.io.Source

case class InputStreamTransactionLoader(source: InputStream) extends TransactionLoader {

  override def load(): List[Transaction] = {
    val reader = Source.fromInputStream(this.source)
    reader.getLines().drop(1).map(buildTransaction).toList
  }

  private def buildTransaction(line: String): Transaction = {
    println("reading line:" + line)
    val fields: Array[String] = line.split(",")
    Transaction(
      fields(0).trim,
      LocalDateTime.parse(fields(1).trim, Constants.dateTimeFormatter),
      BigDecimal(fields(2).trim),
      fields(3).trim,
      TransactionType.withName(fields(4).trim),
      if (fields.length == 6) Some(fields(5).trim) else None
    )
  }

}

File: transaction-analyzer/kotlin/src/main/kotlin/InputStreamTransactionLoader.kt
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.math.BigDecimal
import kotlin.streams.toList

class InputStreamTransactionLoader(val input: InputStream) : TransactionLoader {
    override fun load(): List<Transaction> {
        val reader = BufferedReader(InputStreamReader(input))
        return reader.lines().skip(1).map { buildTransaction(it) }.toList()
    }

    private fun buildTransaction(line: String): Transaction {
        println("build transaction from line: $line")
        val fields = line.split(",")
        return Transaction(
            id = fields[0].trim(),
            transactedAt = fields[1].trim().toLocalDateTime(),
            amount = BigDecimal(fields[2].trim()),
            merchantName = fields[3].trim(),
            type = TransactionType.valueOf(fields[4].trim()),
            relatedTransactionId = fields[5].trim()
        )
    }
}